// @fxts/core/dist/esm/index.js
import add from "./add.js";
import always from "./always.js";
import apply from "./apply.js";
import average from "./average.js";
import compactObject from "./compactObject.js";
import consume from "./consume.js";
import countBy from "./countBy.js";
import curry from "./curry.js";
import delay from "./delay.js";
import each from "./each.js";
import every from "./every.js";
import evolve from "./evolve.js";
import find from "./find.js";
import findIndex from "./findIndex.js";
import fromEntries from "./fromEntries.js";
import groupBy from "./groupBy.js";
import gt from "./gt.js";
import gte from "./gte.js";
import head from "./head.js";
import identity from "./identity.js";
import includes from "./includes.js";
import indexBy from "./indexBy.js";
import isArray from "./isArray.js";
import isBoolean from "./isBoolean.js";
import isEmpty from "./isEmpty.js";
import isNil from "./isNil.js";
import isNull from "./isNull.js";
import isNumber from "./isNumber.js";
import isObject from "./isObject.js";
import isString from "./isString.js";
import isUndefined from "./isUndefined.js";
import join from "./join.js";
import juxt from "./juxt.js";
import last from "./last.js";
import lt from "./lt.js";
import lte from "./lte.js";
import max from "./max.js";
import memoize from "./memoize.js";
import min from "./min.js";
import negate from "./negate.js";
import noop from "./noop.js";
import not from "./not.js";
import nth from "./nth.js";
import omit from "./omit.js";
import omitBy from "./omitBy.js";
import partition from "./partition.js";
import pick from "./pick.js";
import pickBy from "./pickBy.js";
import pipe from "./pipe.js";
import pipe1 from "./pipe1.js";
import prop from "./prop.js";
import props from "./props.js";
import reduce from "./reduce.js";
import size from "./size.js";
import some from "./some.js";
import sort from "./sort.js";
import sortBy from "./sortBy.js";
import sum from "./sum.js";
import tap from "./tap.js";
import throwError from "./throwError.js";
import throwIf from "./throwIf.js";
import toArray from "./toArray.js";
import unicodeToArray from "./unicodeToArray.js";
import unless from "./unless.js";
export { add, always, apply, average, average as mean, compactObject, consume, countBy, curry, delay, each, every, evolve, find, findIndex, fromEntries, groupBy, gt, gte, head, head as first, identity, includes, includes as contains, indexBy, isArray, isBoolean, isEmpty, isNil, isNumber, isObject, isString, isUndefined, isNull, join, juxt, last, lt, lte, max, memoize, min, negate, noop, not, nth, omit, omitBy, partition, pick, pickBy, pipe, pipe1, prop, props, reduce, size, unless, some, sort, sortBy, sum, tap, throwError, throwIf, toArray, unicodeToArray, };
export * from "./Lazy/index.js";
